### ISSUE TITLE
Quest v0.1 / G6PD: Publisher based Questionnaire sync

### ISSUE THREAD

**Issue by pld** [opensrp/fhircore#571]
**Is your feature request related to a problem? Please describe.**
As a developer, I want to be able to sync and trigger Questionnaires based on a variable assigned to the logged in user and variables assigned to Questionnaires.

**Describe the solution you'd like**
- Assign a `publisher` to users (eg somewhere in the user profile we list their `publisher`) and then when that user logs in we sync any Questionnaires that have the same `publisher`, so we’re using the `publisher` to define the “app”.
  - Start with a "g6pd" `publisher` and a user with "g6pd" in their profile.
- Assign `category` to describe the function of the Questionnaire
  - To start we can have a “registration” `category`, see #567 

**Describe alternatives you've considered**
- Sync all Questionnaires for everyone
- Use the `context` field in Questionnaire, doesn't seem as good a fit
- Hard code IDs (what we currently do)

**Additional context**
- See #567 for how we will use the `category` field

**Comment by maimoonak** [comment]
From Mobile and FHIR protocol stand point, yes we can filter Questionnaires by context, publisher, jurisdiction, profile, tag etc ([see others](https://www.hl7.org/fhir/questionnaire.html#search)). 
However, we need to see how user/practitioner can be assigned these params and send with profile details from server. @dubdabasoduba . If any of these are sent with profile we can use and send that as filter for sync/search for Questionnaire

**Comment by trevorgowing** [comment]
Hi @maimoonak. I would like to support with this issue if I can.

You seem to have a good grip on the requirements.

I will appreciate a quick intro to the fhircore app and a briefing on the issue if you have some time this afternoon or tomorrow morning.

**Comment by trevorgowing** [comment]
@pld @f-odhiambo So @maimoonak very graciously gave me a bit of run through of what is required here.

What we discussed was a incremental solution:

1. Filter Questionnaires by `publisher=g6pd` hard-coded.
2. Persist user publisher in shares preferences.
3. Filter Questionnaire by persisted user publisher preference.

Please share any comments or thoughts.

@f-odhiambo Please would you assist in connecting me with the team members tackling the server component of this task so that I can collaborate with them?

**Comment by pld** [comment]
That sounds reasonable. Do you have any thoughts on using `category` to manager where the app should load a specific Questionnaire?

**Comment by trevorgowing** [comment]
> That sounds reasonable. Do you have any thoughts on using `category` to manager where the app should load a specific Questionnaire?

@pld Unless I am missing something, [Questionnaire](http://hl7.org/fhir/questionnaire.html) doesn't have a category field.

Do you not think that perhaps identifier would be a better fit?

**Comment by pld** [comment]
Ah good point, @maimoonak did I mishear you, I thought you mentioned another field to hold info in on the call?

I wonder what the best place to store additional metadata that represents where in a workflow to load the Questionnaire? Identifier certainly seems like a contender, looking at the resource, I've listed some field options below:

1. identifier (multiple values)
2. publisher (single value)
3. useContext (multiple values)
4. code (multiple values)

In the longer-term I could image not wanting to use publisher, since it only stores a single value it won't give us flexibility if we want to define multiple more granular attributes for how to integrate a Questionnaire into the app. I could also imaging publishers, like the WHO, etc wanting to put their name in the publisher field.

[`useContext`](https://hl7.org/fhir/metadatatypes.html#UsageContext) could be nice, in that we could store an Organization (or Group) as a valueReference and then link users to Questionnaires by putting them in the same Org (or Group). But I'm not sure how we'd use it to specify where in the workflow to load the Questionnaire, but I guess a stripped down HealthcareService might work? For example, the patient registration for for G6PD might have a `useContext` value of:

```
[
  {
    code: 'app'
    value:
      {
        valueReference: [the G6PD organization]
      }
  },
  {
    code: 'function`
    value:
      {
        valueReference: [a HealthcareService with only a name field with value "registration"]
      }
  }
]
```

any other suggestions here?

**Comment by dubdabasoduba** [comment]
> From Mobile and FHIR protocol stand point, yes we can filter Questionnaires by context, publisher, jurisdiction, profile, tag etc ([see others](https://www.hl7.org/fhir/questionnaire.html#search)). However, we need to see how user/practitioner can be assigned these params and send with profile details from server. @dubdabasoduba . If any of these are sent with profile we can use and send that as filter for sync/search for Questionnaire

https://github.com/opensrp/web/issues/585 This describes the current practitioner payload that is created from the web.

**Comment by trevorgowing** [comment]
So we had a conversation including @f-odhiambo @ndegwamartin @maimoonak @dubdabasoduba on how to represent the publisher on the server. Our conclusion was to use user roles in keycloak. @dubdabasoduba Did raise some concerns about the scalability of the approach in the context of the single apk approach and requested that some additional thinking been done on this.

So the implementation that I am proceeding with for now will be to extract the publisher from the user roles and filter the Questionnaires based on the value of that publisher role.

Please let us know if there are any questions or concerns.

**Comment by trevorgowing** [comment]
> Ah good point, @maimoonak did I mishear you, I thought you mentioned another field to hold info in on the call?
> 
> I wonder what the best place to store additional metadata that represents where in a workflow to load the Questionnaire? Identifier certainly seems like a contender, looking at the resource, I've listed some field options below:
> 
>     1. identifier (multiple values)
>     2. publisher (single value)
>     3. useContext (multiple values)
>     4. code (multiple values)
> 
> 
> In the longer-term I could image not wanting to use publisher, since it only stores a single value it won't give us flexibility if we want to define multiple more granular attributes for how to integrate a Questionnaire into the app. I could also imaging publishers, like the WHO, etc wanting to put their name in the publisher field.
> 
> [`useContext`](https://hl7.org/fhir/metadatatypes.html#UsageContext) could be nice, in that we could store an Organization (or Group) as a valueReference and then link users to Questionnaires by putting them in the same Org (or Group). But I'm not sure how we'd use it to specify where in the workflow to load the Questionnaire, but I guess a stripped down HealthcareService might work? For example, the patient registration for for G6PD might have a `useContext` value of:
> 
> ```
> [
>   {
>     code: 'app'
>     value:
>       {
>         valueReference: [the G6PD organization]
>       }
>   },
>   {
>     code: 'function`
>     value:
>       {
>         valueReference: [a HealthcareService with only a name field with value "registration"]
>       }
>   }
> ]
> ```
> 
> any other suggestions here?

@pld @maimoonak If we are using the publisher to represent the author/organization that created/designed the Questionnaire which I think it the right way to go, I don't think that it should be used to represent the stage in the workflow. I think that [`useContext`](https://hl7.org/fhir/metadatatypes.html#UsageContext) sounds like the correct field to me.

**Comment by ndegwamartin** [comment]
@dubdabasoduba for the concern you raised earlier about alignment of the _fhircore_ vs _quest_ auth flow for questionnaires in relation to the Single APK approach, I think we can solve it by only syncing down questionnaires based on the module configuration . i.e. at the point where we have the configurations for which modules _(anc, eir e.t.c)_ the specific deployment should have, we should also have configurations for the specific (Questionnaire) resources they can sync down. 

Alternatively we could somehow extrapolate that from the configured modules and figure out which questionnaires to sync down to the device. I'm thinking this would be at installation and also subsequent syncs as assignments might be dynamic

Let me know your thoughts on this

**Comment by ndegwamartin** [comment]
I guess another important question would be do we envision a situation where the **quest mode** and the **Single APK mode** are not mutually exclusive? cc @pld @f-odhiambo

**Comment by ndegwamartin** [comment]
I guess another important question would be, do we envision a situation where the **Quest mode** and the **Single APK mode** _are not_ mutually exclusive? cc @pld @f-odhiambo

**Comment by pld** [comment]
@ndegwamartin yes I think that is possible but we don't enough information, we discussed this a bit in our Quest architecture call, I see the work we're doing now as a way to see if Quest can be the single app, if we can get it configurable enough to be that.

@trevorgowing @dubdabasoduba I also think we should go with `useContext` to keep `publisher` available for it's semantically appropriate values -- we can delay that switch if we're far enough down the path that this adds too much work

**Comment by trevorgowing** [comment]
@pld @ndegwamartin @dubdabasoduba Just a heads up. We had to configure keycloak to expose the roles in the `userInfo` endpoint. We have done that now for the `FHIR_Android` realm as you can see below.

```json
{
    "sub": "dbde5c0c-2aa0-4278-a505-f6ef9d262a1b",
    "email_verified": false,
    "realm_access": {
        "roles": [
            "g6pd-questionnaire",
            "offline_access",
            "default-roles-fhir_android",
            "uma_authorization",
            "g6pd"
        ]
    },
    "name": "G6PD Test",
    "groups": [
        "g6pd-questionnaire",
        "offline_access",
        "default-roles-fhir_android",
        "uma_authorization",
        "g6pd"
    ],
    "preferred_username": "g6pd",
    "given_name": "G6PD",
    "family_name": "Test",
    "email": "g6pddemo@ona.io"
}
```

**Comment by trevorgowing** [comment]
@f-odhiambo I can't see how any UI would relate to this issue.

I would expect that a list of Questionnaires would be displayed, all I am doing is filtering the list.

**Comment by pld** [comment]
Agree, any UI for this is handled by #585

**Comment by trevorgowing** [comment]
@syedowaisali90 Please feel free to reach out if you have any questions etc.

**Comment by f-odhiambo** [comment]
I meant to say he will need it 😊 to view synced questionnaires

**Comment by trevorgowing** [comment]
[PR 590](https://github.com/opensrp/fhircore/pull/590) Ready for Review.
